bleak/backends/bluezdbus/agent.py

```diff
@@ -0,0 +1,210 @@
+"""
+Agent
+-----
+
+This module contains types associated with the BlueZ D-Bus `agent api
+<https://github.com/bluez/bluez/blob/master/doc/agent-api.txt>`.
+"""
+
+import asyncio
+import contextlib
+import logging
+import os
+from typing import Set, no_type_check
+
+from dbus_fast import DBusError, Message
+from dbus_fast.aio import MessageBus
+from dbus_fast.service import ServiceInterface, method
+from dbus_fast.signature import Variant
+
+from bleak.agent import BaseBleakAgentCallbacks
+from bleak.backends.bluezdbus import defs
+from bleak.backends.bluezdbus.manager import get_global_bluez_manager
+from bleak.backends.bluezdbus.utils import assert_reply
+from bleak.backends.device import BLEDevice
+
+logger = logging.getLogger(__name__)
+
+
+class Agent(ServiceInterface):
+    """
+    Implementation of the org.bluez.Agent1 D-Bus interface.
+    """
+
+    def __init__(self, callbacks: BaseBleakAgentCallbacks):
+        """
+        Args:
+        """
+        super().__init__(defs.AGENT_INTERFACE)
+        self._callbacks = callbacks
+        self._tasks: Set[asyncio.Task] = set()
+
+    @staticmethod
+    async def _create_ble_device(device_path: str) -> BLEDevice:
+        manager = await get_global_bluez_manager()
+        return BLEDevice(
+            manager.get_device_address(device_path),
+            manager.get_device_name(device_path),
+            {"path": device_path},
+        )
+
+    @method()
+    def Release(self):  # noqa: N802
+        logger.debug("Release")
+
+    # REVISIT: mypy is broke, so we have to add redundant @no_type_check
+    # https://github.com/python/mypy/issues/6583
+
+    @method()
+    @no_type_check
+    async def RequestPinCode(self, device: "o") -> "s":  # noqa: F821 N802
+        logger.debug("RequestPinCode %s", device)
+        raise NotImplementedError
+
+    @method()
+    @no_type_check
+    async def DisplayPinCode(self, device: "o", pincode: "s"):  # noqa: F821 N802
+        logger.debug("DisplayPinCode %s %s", device, pincode)
+        raise NotImplementedError
+
+    @method()
+    @no_type_check
+    async def RequestPasskey(self, device: "o") -> "u":  # noqa: F821 N802
+        logger.debug("RequestPasskey %s", device)
+
+        ble_device = await self._create_ble_device(device)
```

Comment:
Hmm... this is slightly dangerous because "Cancel" could be called if this actually awaits. In that case we would miss the cancellation.

Although the only time this should actually do that is if we lost the D-Bus connection, which means we have bigger problems anyway.

I suppose just adding a comment explaining this would suffice.

Alternative could be to pass `manager` to the constructor so that we have a reference to use without awaiting.